tests/cpp/transforms/make_block_local_test.cpp

```diff
@@ -0,0 +1,196 @@
+    //
+    // |root_snode_|  // ti.root
+    // \
+    //  |
+    //  +- |pointer_snode_|  // .pointer(ti.ij, |pointer_size|)
+    //     \
+    //      |
+    //      +- |bls_snode_|  // .dense(ti.ij, |block_size|s)
+    //      |
+    //      +- |struct_for_snode_|  // .dynamic(ti.k, ...)
+    //
+    // The |struct_for_snode_| is the one being iterated over in the struct-for
+    // offloaded task |for_stmt_|, meaning that the loop index is based on this
+    // SNode. On the other hand, |bls_snode_| is the one we want to cache into
+    // the BLS buffer (shared memory).
+    //
+    //
+    // |bls_snode_| has a larger shape than |struct_for_snode_|, because we
+    // want to see if the tests can handle the loop index scaling multiplier
+    // (block_size) and infer the BLS size correctly.
+    const std::vector<Index> indices = {Index{0}, Index{1}};
+    pointer_snode_ = &(root_snode_->pointer(indices, pointer_size));
+
+    bls_snode_ = &(pointer_snode_->dense(indices, /*sizes=*/block_size));
+    bls_place_snode_ = &(bls_snode_->insert_children(SNodeType::place));
+    bls_place_snode_->dt = PrimitiveType::f32;
+
+    struct_for_snode_ =
+        &(pointer_snode_->dynamic({Index{2}}, /*n=*/1024, /*chunk_size=*/128));
+    struct_for_place_snode_ =
+        &(struct_for_snode_->insert_children(SNodeType::place));
+    struct_for_place_snode_->dt = PrimitiveType::i32;
+
+    FakeStructCompiler sc;
+    sc.run(*root_snode_, /*unused=*/false);
+
+    for_stmt_ = std::make_unique<OffloadedStmt>(
+        /*task_type=*/OffloadedTaskType::struct_for,
+        /*arch=*/Arch::x64);
+    for_stmt_->mem_access_opt.add_flag(bls_place_snode_,
+                                       SNodeAccessFlag::block_local);
+    for_stmt_->snode = struct_for_place_snode_;
+    for_stmt_->block_dim = 64;
+
+    builder_.set_insertion_point(
+        {/*block=*/for_stmt_->body.get(), /*position=*/0});
+  }
+
+  int get_block_corner(int loop_index, int axis) const {
+    // TODO(#2327): Once we switch to consecutive indexing, we can simply return
+    // |loop_index| itself: For |struct_for_snode_|, the coefficients
+    // along ti.i and ti.j w.r.t. |pointer_snode_| are just 1.
+    unsigned mask = get_block_size(axis) - 1;
+    mask = ~mask;
+    return (loop_index & mask);
+  }
+
+  int get_block_size(int axis) const {
+    axis = bls_snode_->physical_index_position[axis];
+    return (1 << bls_snode_->extractors[axis].num_bits);
+  }
+
+  std::unique_ptr<SNode> root_snode_{nullptr};
+  SNode *pointer_snode_{nullptr};
+  SNode *bls_snode_{nullptr};
+  SNode *bls_place_snode_{nullptr};
+  SNode *struct_for_snode_{nullptr};
+  SNode *struct_for_place_snode_{nullptr};
+  std::unique_ptr<OffloadedStmt> for_stmt_{nullptr};
+
+  IRBuilder builder_;
+};
+
+TEST_F(MakeBlockLocalTest, Basic) {
+  initialize(/*pointer_size=*/2, /*block_size=*/4);
+
+  // This test has two global loads to |bls_place_snode_|:
+  // * [i, j]
```

Comment:
@KLozes In #2327, could you change this to multiply with the block size? I'll help review the changes as well :-)

Reply:
Sure thing!